### ISSUE TITLE
Code Inspection Idea: Identify "Magic Numbers"

### ISSUE THREAD

**Issue by mwolfe02** [retailcoder/Rubberduck#359]
Good programming practice is to assign numbers to constants instead of using them directly in code.  To support this I would recommend the following new code inspection:

Name: Use of "Magic Number"
Type: MaintainabilityAndReadabilityIssues
Severity: Suggestion

Algorithm:

0. Identify all number tokens (e.g., `3`, `1.5`, `255`, `&hFF`)
0. Ignore tokens that are the target of a `Const` statement (e.g., `Const DaysInWeek = 7`)
0. Ignore `0` and `1` tokens

**Comment by ckuhn203** [comment]
I'd like to amend the last one.

3. Ignore `0` and `1` tokens where they occur in a loop declaration.

    For i = 0 To Blah + 1
        ' some code
    Next i

Is this your intent @mwolfe02 , or do you think it's better to exclude all occurrences of `0` and `1`?

**Comment by mwolfe02** [comment]
I think it's better to include all occurrences of 0 and 1.  Consider the following:

    If DCount("*", "SomeTable") > 0 Then

    FName = Dir("C:\*")
    Do While Len(FName) > 0
        FName = Dir()
    Loop

    IsPlural = (MyCollection.Count > 1)

**Comment by ckuhn203** [comment]
Point taken. I was just refrerenced a section I recently read in Code Complete and it would appear Steve McConnel is in agreement with you. That's good enough for me.

**Comment by retailcoder** [comment]
This might be simpler than I initially thought it would be, although it implies walking the tree and listening for `ValueStmt` rules; if the value statement does not have a `ConstSubStmtContext` parent then the magic numbers are any number literal greater than 1. Right?

**Comment by retailcoder** [comment]
The main problem is that `ValueStmt` is all over the grammar, and all over the code - this inspection may cause performance issues whenever its severity is set greater than `DoNotShow`.

**Comment by mwolfe02** [comment]
@retailcoder Actually, any number literal greater than 1 or less than 0.

**Comment by ckuhn203** [comment]
:+1: Good catch.

**Comment by Zomis** [comment]
Related: [Is 100 a magic number?](http://codereview.stackexchange.com/q/44047/31562)

**Comment by retailcoder** [comment]
@Zomis @ckuhn203 should we have a configurable "list of acceptable magic numbers"?

**Comment by ckuhn203** [comment]
Yes @retailcoder. We should. We can put `0` & `1` in there as defaults, but it should be configurable. It will be very similar to the code for the configurable `TodoMarkers`.

**Comment by Hosch250** [comment]
Well, configurable or not, we shouldn't directly alert about magic numbers for `Option Base 1`.

**Comment by ThunderFrame** [comment]
linking #3055
